eth-contracts/contracts/Governance.sol

```diff
@@ -4,284 +4,350 @@ import "./service/registry/RegistryContract.sol";
-    address proposer,
-    uint256 startBlockNumber,
-    address target,
-    uint256 slashAmount,
-    Outcome outcome,
-    uint256 voteMagnitudeYes,
-    uint256 voteMagnitudeNo,
-    uint256 numVotes
-  ) {
-    require(_proposalId <= lastProposalId && _proposalId > 0, "Must provide valid non-zero _proposalId");
-
-    SlashProposal memory proposal = slashProposals[_proposalId];
-    return (
-      proposal.proposalId,
-      proposal.proposer,
-      proposal.startBlockNumber,
-      proposal.target,
-      proposal.slashAmount,
-      proposal.outcome,
-      proposal.voteMagnitudeYes,
-      proposal.voteMagnitudeNo,
-      proposal.numVotes
-      /** @notice - votes mapping cannot be returned by external function */
-    );
-  }
+    // ========================================= Getters =========================================
+
+    function getProposalById(uint256 _proposalId)
+    external view returns (
+        uint256 proposalId,
+        address proposer,
+        uint256 startBlockNumber,
+        address targetContract,
+        uint callValue,
+        string memory signature,
+        bytes memory callData,
+        Outcome outcome,
+        uint256 voteMagnitudeYes,
+        uint256 voteMagnitudeNo,
+        uint256 numVotes
+    )
+    {
+        require(
+            _proposalId <= lastProposalId && _proposalId > 0,
+            "Must provide valid non-zero _proposalId"
+        );
+
+        Proposal memory proposal = proposals[_proposalId];
+        return (
+            proposal.proposalId,
+            proposal.proposer,
+            proposal.startBlockNumber,
+            proposal.targetContract,
+            proposal.callValue,
+            proposal.signature,
+            proposal.callData,
+            proposal.outcome,
+            proposal.voteMagnitudeYes,
+            proposal.voteMagnitudeNo,
+            proposal.numVotes
+            /** @notice - votes mapping cannot be returned by external function */
+        );
+    }
 
-  function getVoteBySlashProposalAndVoter(uint256 _proposalId, address _voter)
-  external view returns (Vote vote)
-  {
-    require(_proposalId <= lastProposalId && _proposalId > 0, "Must provide valid non-zero _proposalId");
-    return slashProposals[_proposalId].votes[_voter];
-  }
+    function getVoteByProposalAndVoter(uint256 _proposalId, address _voter)
+    external view returns (Vote vote)
+    {
+        require(
+            _proposalId <= lastProposalId && _proposalId > 0,
+            "Must provide valid non-zero _proposalId"
+        );
+        return proposals[_proposalId].votes[_voter];
+    }
 
+    // ========================================= Private =========================================
+
+    function _executeTransaction(
+        address _targetContract,
+        uint256 _callValue,
+        string memory _signature,
+        bytes memory _callData
+    ) internal returns (bytes memory /** returnData */)
+    {
+        bytes32 txHash = keccak256(
```

Comment:
can this ever be duplicated? lets say a proposer re-submits the same proposal w/everything equal

Reply:
oh it definitely can be duplicated, but it's not used anywhere onchain. only used offchain to ensure that the expected data was called. not a uniqueness enforcement